FWCore/Utilities/interface/propagate_const_array.h

```diff
@@ -0,0 +1,133 @@
+
+    template <typename T>
+    constexpr auto has_subscript_operator_v = has_subscript_operator<T>::value;
+
+    // for a type T, return the type of the return value of the subscript operator T[N]
+    template <typename T, typename = void, typename = void>
+    struct subscript_type {};
+
+    // the specialisations for arrays allow supporting incomplete types
+    template <typename T>
+    struct subscript_type<T[]> {
+      using type = T;
+    };
+
+    template <typename T, int N>
+    struct subscript_type<T[N]> {
+      using type = T;
+    };
+
+    // for non-array types that implement the subscript operator[], a complete type is needed
+    template <typename T>
+    struct subscript_type<T, std::enable_if_t<not std::is_array_v<T>>, std::enable_if_t<has_subscript_operator_v<T>>> {
+      using type = typename std::remove_reference<decltype(std::declval<T&>()[0])>::type;
+    };
+
+    template <typename T>
+    using subscript_type_t = typename subscript_type<T>::type;
+
+  }  // namespace impl
+
+  template <typename T>
+  class propagate_const_array;
+
+  template <typename T>
+  std::decay_t<T>& get_underlying(propagate_const_array<T>&);
+  template <typename T>
+  std::decay_t<T> const& get_underlying(propagate_const_array<T> const&);
+
+  template <typename T>
+  class propagate_const_array {
+  public:
+    friend std::decay_t<T>& get_underlying<T>(propagate_const_array<T>&);
+    friend std::decay_t<T> const& get_underlying<T>(propagate_const_array<T> const&);
+
+    template <typename U>
+    friend class propagate_const_array;
+
+    //using element_type = typename std::remove_reference<decltype(std::declval<T&>()[0])>::type;
+    using element_type = typename impl::subscript_type_t<T>;
+
+    propagate_const_array() = default;
+
+    propagate_const_array(propagate_const_array<T>&&) = default;
+    propagate_const_array(propagate_const_array<T> const&) = delete;
+    template <typename U>
+    propagate_const_array(U&& iValue) : m_value(std::forward<U>(iValue)) {}
+
+    propagate_const_array<T>& operator=(propagate_const_array&&) = default;
+    propagate_const_array<T>& operator=(propagate_const_array<T> const&) = delete;
+
+    template <typename U>
+    propagate_const_array& operator=(propagate_const_array<U>& other) {
+      static_assert(std::is_convertible_v<std::decay_t<U>, std::decay_t<T>>,
+                    "Cannot assign propagate_const_array<> of incompatible types");
+      m_value = other.m_value;
+      return *this;
+    }
+
+    template <typename U>
+    propagate_const_array& operator=(U&& iValue) {
+      m_value = std::forward<U>(iValue);
+      return *this;
+    }
+
+    // ---------- const member functions ---------------------
+    element_type const* get() const { return &m_value[0]; }
+    element_type const& at(size_t pos) const { return m_value[pos]; }
+    element_type const& operator[](size_t pos) const { return this->at(pos); }
```

Comment:
Are the implementations possibly reversed?
```suggestion
    element_type const& at(size_t pos) const { return this->at(pos); }
    element_type const& operator[](size_t pos) const { return m_value[pos]; }
```

Reply:
The suggested change would make `at(pos)` an infinite recursion.

The idea was to mirror the interface of `propagate_const<T>::operator*` and `propagate_const<T>::get()`:
  - the underlying type may not have `at()` (like the underlying ype of `propagate_const<T>` may not have `get()`)
  - `at(pos)` implements access to the `pos` element (like `get()` returns the underlying pointer)
  - `operator[](pos)` is equivalent to `at(pos)` (like `operator*()` is equivalent to `get()`)

Reply:
Oh yeah, didn't think it through. I'm a bit worried that using `at()` could create confusion because `std` containers do bounds check with `at()`.

> * `operator[](pos)` is equivalent to `at(pos)` (like `operator*()` is equivalent to `get()`)

Should `at(pos)` then return a pointer?

`std::unique_ptr<T[]>` does not have "an equivalent method of `operator[](pos)` like `operator*()` is equivalent to `get()`", so we don't have a naming example from `std` for this kind of functionality.

How about calling it `getAt(post)` or similar?

Reply:
Following Chris comment I decided to simple remove it.